[quant][fx] Don't assume bias is a keyword argument

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61646
* #61608
* __->__ #61647

`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a keyword argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

Differential Revision: [D31401360](https://our.internmc.facebook.com/intern/diff/D31401360)

peterbell10 committed:
[quant][fx] Don't assume bias is a keyword argument

`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a key argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

[ghstack-poisoned]

peterbell10 committed:
Update on "[quant][fx] Don't assume bias is a keyword argument"


`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a key argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

[ghstack-poisoned]

peterbell10 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -754,10 +754,8 @@ def convert(self,
                         bias = load_arg(quantized=torch.float)(self.linear_node.args[2])
                         other_args = other_args[1:]  # remove the bias argument
                     else:
-                        assert 'bias' in kwargs, \
-                            'expect bias provided as a keyword argument when it is not a positional argument'
```
This also only works coincidentally. `bias` is an optional argument, but for python bound methods defaults are currently handled before `__torch_function__` but there's no guaruntee this is the case, and this isn't true for C++ bound functions.

mruberry commented:
@jamesr66a would you review this or suggest another reviewer?

peterbell10 committed:
Update on "[quant][fx] Don't assume bias is a keyword argument"


`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a key argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

[ghstack-poisoned]

peterbell10 committed:
Update on "[quant][fx] Don't assume bias is a keyword argument"


`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a key argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

[ghstack-poisoned]

peterbell10 commented:
@jamesr66a ping.

peterbell10 commented:
@mruberry are there any active fx reviewers?

mruberry commented:
cc @Chillee, too

Chillee commented:
Not to shirk responsibility :P, but this code is primarily written by the quantization team and not the FX team, I believe. 

cc: @jerryzh168, @z-a-f, @HDCharles, judging by recent commits to the file.

peterbell10 committed:
Update on "[quant][fx] Don't assume bias is a keyword argument"


`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a key argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -104,24 +104,17 @@ def node_arg_is_weight(node: Node, arg: Any) -> bool:
+    if not isinstance(node, Node) or node.op != 'call_function':
+        return False
+
+    assert not isinstance(node.target, str), "expected a function target"
```
this is a check function, should we have assert here?

peterbell10 replied:
I though this was necessary for mypy, but it seems to be passing without it so I guess not.

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -1033,10 +1033,8 @@ def convert(self,
-                            'expect bias provided as a keyword argument when it is not a positional argument'
-                        bias = kwargs['bias']
-                        kwargs.pop('bias')
+                        kwargs.pop('bias', None)
```
can you add a comment in the code here? (maybe copying the comment in the github) I thought we will either have positional or keyword bias arg

peterbell10 replied:
>I thought we will either have positional or keyword bias arg

AFAIK fx traces the arguments as they were passed to `__torch_function__` which for C++ functions will match how the user called it. So as long as the signature allows it, any argument could be either positional or keyword.

jerryzh168 reviewed (approved):
lgtm, thanks!

peterbell10 committed:
Update on "[quant][fx] Don't assume bias is a keyword argument"


`prepare_fx` currently assumes that bias is always a positional argument to
convolutions, and only a keyword argument to other functions. This happens to work
today due to a quirk in how `__torch_function__` is handled for python
functions but shouldn't be considered stable.

Instead, we should support `bias` for both positional and keyword forms.

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31401360).
